### PR TITLE
Issue #2408: Updated warning messages for JavadocMethod

### LINKED ISSUE #2408: JavadocMethod: unclear message for incomplete tag
```
$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.2//EN"
        "http://jee-qa/jee-qa/trunk/checkstyle/configuration_1_2.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="JavadocMethod" />
    </module>
</module>

$ cat Test.java 
class A {
 /**
 * Handle query request
 *
 * @param mpIdParm das
 * @param planYearParm sa
 * @param modelParm sd
 * @return the name of the view
 */
@RequestMapping("/query/mpId/{mpId}/planYear/{planYear}")
public String handleQuery(@PathVariable("mpId") final String mpIdParm,
        @PathVariable("planYear") final String planYearParm, final Model modelParm) {
 }
}
rivanov@p5510:/var/tmp$ java -jar /var/tmp/checkstyle-8.32-all.jar \
      -c config.xml Test.java 
Starting audit...
Audit done.

# NO RETURN TAG 
$cat Test.java 
class A {
 /**
 * Handle query request
 *
 * @param mpIdParm das
 * @param planYearParm sa
 * @param modelParm sd
 */
@RequestMapping("/query/mpId/{mpId}/planYear/{planYear}")
public String handleQuery(@PathVariable("mpId") final String mpIdParm,
        @PathVariable("planYear") final String planYearParm, final Model modelParm) {
 }
}

$ java $RUN_LOCALE -jar /var/tmp/checkstyle-8.33-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /var/tmp/Test.java:9: Expected @return tag. [JavadocMethod]
Audit done.
Checkstyle ends with 1 errors.

# EMPTY RETURN TAG
$ cat Test.java 
class A {
 /**
 * Handle query request
 *
 * @param mpIdParm
 * @param planYearParm
 * @param modelParm
 * @return
 */
@RequestMapping("/query/mpId/{mpId}/planYear/{planYear}")
public String handleQuery(@PathVariable("mpId") final String mpIdParm,
        @PathVariable("planYear") final String planYearParm, final Model modelParm) {
 }
}
rivanov@p5510:/var/tmp$ java -jar /var/tmp/checkstyle-8.32-all.jar \
     -c config.xml Test.java 
Starting audit...
[ERROR] /var/tmp/Test.java:10: Expected @return tag. [JavadocMethod]
Audit done.
Checkstyle ends with 1 errors.
```

Expected message:
[ERROR] /var/tmp/Test.java:10: @return tag should present and have description. [JavadocMethod]
##

### PR DESCRIPTION
**Pull request by shashwatj07**
Resolves #2408

### CODE REVIEW

**Comment by rnveach** [comment]
@shashwat70 There is no changes needed to the logic of the check? They work as intended and the first part of the issue in the description isn't a concern anymore? 

Please provide us with CLI examples using latest version of check.

**Comment by rnveach** [comment]
Also message update is a general message used for many different tags. Please provide a list of tags and examples for each one to show we are not damaging other tags.

**Comment by shashwat70** [comment]
Hear it is! @rnveach 

```
C:\Users\Shashwat\Documents>cat Test.java
class A {
    /**
    * Handle query request
    *
    * @param mpIdParm das
    * @param planYearParm sa
    * @param modelParm sd
    * @return the name of the view
    */
   @RequestMapping("/query/mpId/{mpId}/planYear/{planYear}")
   public String handleQuery(@PathVariable("mpId") final String mpIdParm,
           @PathVariable("planYear") final String planYearParm, final Model modelParm) {
    }
   }
C:\Users\Shashwat\Documents>cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="JavadocMethod">
    </module>
  </module>
</module>
C:\Users\Shashwat\Documents>java -jar checkstyle-8.31-all.jar -c config.xml Test.java
Starting audit...
Audit done.
```

```
C:\Users\Shashwat\Documents>cat Test.java
class A {
    /**
    * Handle query request
    *
    * @param planYearParm sa
    * @param modelParm sd
    * @return the name of the view
    */
   @RequestMapping("/query/mpId/{mpId}/planYear/{planYear}")
   public String handleQuery(@PathVariable("mpId") final String mpIdParm,
           @PathVariable("planYear") final String planYearParm, final Model modelParm) {
    }
   }
C:\Users\Shashwat\Documents>cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="JavadocMethod">
    </module>
  </module>
</module>
C:\Users\Shashwat\Documents>java -jar checkstyle-8.31-all.jar -c config.xml Test.java
Starting audit...
[ERROR] C:\Users\Shashwat\Documents\Test.java:10:65: Expected @param tag for 'mpIdParm'. [JavadocMethod]
Audit done.
Checkstyle ends with 1 errors.
```

**Comment by shashwat70** [comment]
> Please provide a list of tags and examples for each one to show we are not damaging other tags.

I'm sorry but I'm unable to understand what you mean by this. Could you please elaborate?

**Comment by baratali** [comment]
https://github.com/checkstyle/checkstyle/pull/8071#issuecomment-612463481
@shashwat70 , can you please test the same examples with checkstyle from your branch (including this commit)? We need to see new violation message in CLI

----

> > Please provide a list of tags and examples for each one to show we are not damaging other tags.
> 
> I'm sorry but I'm unable to understand what you mean by this. Could you please elaborate?

You changed javadoc.expectedTag message and it could be used for other tags as well (not only `param`).
Please find for what tags javadoc.expectedTag message is used.
Try to track this constant:
```
src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java:    public static final String MSG_EXPECTED_TAG = "javadoc.expectedTag";
```
If this message is used to print warnings for other tags, please provide CLI example for each one.

**Comment by shashwat70** [comment]
> Please find for what tags ```javadoc.expectedTag``` message is used.

this is used for ```@throws``` and ```@param``` tags.

**Comment by shashwat70** [comment]
> If this message is used to print warnings for other tags, please provide CLI example for each one.

please refer https://github.com/checkstyle/checkstyle/pull/8071#issuecomment-612463481
I've updated it. @baratali

**Comment by shashwat70** [comment]
I believe the issue should also extend to ```@return``` tag.

```
C:\Users\Shashwat\Documents>cat Test.java
class A {
    /**
    * Handle query request
    *
    * @param mpIdParm abcd
    * @param planYearParm sa
    * @param modelParm sd
    * @throws IOException abcd
    */
   @RequestMapping("/query/mpId/{mpId}/planYear/{planYear}")
   public String handleQuery(@PathVariable("mpId") final String mpIdParm,
           @PathVariable("planYear") final String planYearParm, final Model modelParm) throws IOException{
    }
   }
C:\Users\Shashwat\Documents>java -jar checkstyle-8.32-SNAPSHOT-all.jar -c config.xml Test.java
Starting audit...
[ERROR] C:\Users\Shashwat\Documents\Test.java:10: Expected @return tag. [JavadocMethod]
Audit done.
Checkstyle ends with 1 errors.
```

A more self-explanatory message would be:
```Test.java:10: Expected @return tag with description. [JavadocMethod]```
(The check expects description too)

**Comment by baratali** [comment]
Hey. I have tested the Check and it doesn't actually require a description for `@param` and `@throws`.
Looks like Check's logic changed since https://github.com/checkstyle/checkstyle/issues/2408 was created in 2015.
However, `@return` requires a description.

Here is my test.
```xml
    <module name="JavadocMethod">
      <property name="validateThrows" value="true"/>
      <property name="allowMissingParamTags" value="false"/>
      <property name="allowMissingReturnTag" value="false"/>
    </module>
```
```java
class Input {
	/**
	 * test1
	 */
	public <T> int test1(T arg1, Integer arg2) throws IOException {
		return 1;
	}

	/**
	 * test2
	 * @param arg1
	 * @param arg2
	 * @param <T>
	 * @throws IOException
	 * @return
	 */
	public <T> int test2(T arg1, Integer arg2) throws IOException {
		return 1;
	}

	/**
	 * test3
	 * @param arg1 text
	 * @param arg2 text
	 * @param <T> text
	 * @throws IOException text
	 * @return text
	 */
	public <T> int test3(T arg1, Integer arg2) throws IOException {
		return 1;
	}
}
```
```
$ java -jar checkstyle-8.33-SNAPSHOT-all.jar -c config.xml Input.java
Starting audit...
[ERROR] C:\Users\Baratali\git\test_checkstyle\2\Input.java:5: Expected @return tag. [JavadocMethod]
[ERROR] C:\Users\Baratali\git\test_checkstyle\2\Input.java:5:17: Expected @param tag for '<T>'. [JavadocMethod]
[ERROR] C:\Users\Baratali\git\test_checkstyle\2\Input.java:5:32: Expected @param tag for 'arg1'. [JavadocMethod]
[ERROR] C:\Users\Baratali\git\test_checkstyle\2\Input.java:5:46: Expected @param tag for 'arg2'. [JavadocMethod]
[ERROR] C:\Users\Baratali\git\test_checkstyle\2\Input.java:5:59: Expected @throws tag for 'IOException'. [JavadocMethod]
[ERROR] C:\Users\Baratali\git\test_checkstyle\2\Input.java:17: Expected @return tag. [JavadocMethod]
Audit done.
Checkstyle ends with 6 errors.
```
As you can see the Check doesn't warn on params and exception in test2(). It only warns about return tag.

@shashwat70 , please keep changes only for `@return` message and remove changes for `@param`/`@throws`.
Probably the Check should verify a description, but it's another issue.

**Comment by shashwat70** [comment]
> please keep changes only for @return message and remove changes for @param/@throws.

done @baratali

**Comment by baratali** [comment]
@shashwat70 , please change commit message too.
Also please rebase your branch on latest master from checkstyle/checkstyle.

**Comment by shashwat70** [comment]
> @shashwat70 , please change commit message too.
> Also please rebase your branch on latest master from checkstyle/checkstyle.

Done @baratali

**Comment by baratali** [comment]
@romani , please review

**Comment by baratali** [comment]
From https://github.com/checkstyle/checkstyle/pull/8071#issuecomment-625545952:

> Probably the Check should verify a description for @param/@throws, but it's another issue.

Such issue already exist - https://github.com/checkstyle/checkstyle/issues/6746. I've put a comment about `@throws` tag as well: https://github.com/checkstyle/checkstyle/issues/6746#issuecomment-627605414

File: src/main/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/messages.properties [hunk 1]
```diff
@@ -21,7 +21,7 @@ javadoc.paragraph.misplaced.tag=<p> tag should be placed immediately before the
 javadoc.paragraph.redundant.paragraph=Redundant <p> tag.
 javadoc.paragraph.tag.after=Empty line should be followed by <p> tag on the next line.
 javadoc.parse.rule.error=Javadoc comment at column {0} has parse error. Details: {1} while parsing {2}
-javadoc.return.expected=Expected @return tag.
+javadoc.return.expected=Expected @return tag with description.
```
Comment (romani) [inline src/main/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/messages.properties:24]: In issue there is a request  to update message
https://github.com/checkstyle/checkstyle/blob/1cc84e8b8f052249b10a11400a5571cfdaa30e21/src/main/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/messages.properties#L9

please update it also.

Comment (romani) [reply src/main/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/messages.properties:24]: I updated issue description with ltest behavior, Check do not report problems with such message.

Comment (romani) [inline src/main/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/messages.properties:24]: I updated expected message at issue description, sorry for confusion, please do.

**Comment by romani** [comment]
@shashwatj07 , ping.

File: src/main/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/messages.properties (hunk 1 above)
Comment (shashwatj07) [reply src/main/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/messages.properties]: done

**Comment by shashwatj07** [comment]
> @shashwatj07 , ping.

Please review. @romani

File: src/main/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/messages.properties [hunk 2]
```diff
@@ -21,7 +21,7 @@ javadoc.paragraph.misplaced.tag=<p> tag should be placed immediately before the
 javadoc.paragraph.redundant.paragraph=Redundant <p> tag.
 javadoc.paragraph.tag.after=Empty line should be followed by <p> tag on the next line.
 javadoc.parse.rule.error=Javadoc comment at column {0} has parse error. Details: {1} while parsing {2}
-javadoc.return.expected=Expected @return tag.
+javadoc.return.expected=@return tag should have description.
```
Comment (romani) [inline src/main/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/messages.properties:24]: `@return tag should be present and have description.`

Comment (romani) [reply src/main/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/messages.properties:24]: @ shashwatj07 , please do.
Sorry for several confusions.

Comment (shashwatj07) [reply src/main/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/messages.properties]: Done @romani

**Pull request merged**